Add hovered block store and show hovered template part's label in edit site header component.

Addison-Stavlo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

Adds a way to track hovered blocks by adding them as an array of clientIds in the `block-editor` store.  This information is then used to show the hovered template part's label in the component added in #25320 (alternative approaches discussed in #25469).

![hovered-template-part-label](https://user-images.githubusercontent.com/28742426/93951844-7195ba80-fd15-11ea-852f-46347ab10341.gif)

The first approach to implement this store tried to add/remove blocks individually by clientId.  However, this ran into a blocker with the inserter triggering removal for all blocks that it appeared to be within the bounds of.  Parsing the blocks found under the cursor and setting the full list solves this problem, and the hovered name is no longer disrupted by the inserter as shown in the gif above. 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
* Tested in the Site Editor
   * hover over a template part and verify its label is displayed in the middle of the top toolbar.
   * hover over a block inserter within the template part and verify the label in the top toolbar does not disappear.
* Smoke tested in Post Editor.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
add basic store

Addison-Stavlo committed:
add mouseEnter/Leave on wrapper div for now...

Addison-Stavlo committed:
basic display name logic, ugly second header to see it

Addison-Stavlo commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -168,19 +171,24 @@ function BlockListBlock( {
-			isSelectionEnabled={ isSelectionEnabled }
-			toggleSelection={ toggleSelection }
-		/>
+		<div
+			onMouseEnter={ () => addHoveredBlock( clientId ) }
+			onMouseLeave={ () => removeHoveredBlock( clientId ) }
+		>
```
We will probably want a better place to add these than tossing in a new div wrapper.  These seem to get overridden if they are placed in the BlockEdit component below.

noahtallen replied:
if there are existing handlers for mouse enter/leave, should we tie into those possibly?

Addison-Stavlo replied:
definitely!  I was getting a little lost trying to find those and put it here for now.  😆

Addison-Stavlo commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -79,64 +80,92 @@ export default function Header( {
+			( blockId ) => getBlockName( blockId ) === 'core/template-part'
+		);
+		const templatePartBlock = getBlock( templatePartBlockId );
+		return templatePartBlock?.attributes.slug || '';
```
Will want to update this to get the actual Title from the entity if available.

noahtallen replied:
we can probably use `getBlockLabel` for this (we use it in other parts of the interface that say "header"). 

but we can tie that in once we are working with the other component :)

Addison-Stavlo commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -79,64 +80,92 @@ export default function Header( {
-						'Add block',
-						'Generic label for block inserter button'
+		<>
+			<div className="edit-site-header">{ nameToDisplay }</div>
```
ugly temporary thing so I could show the hovered name somewhere atm 😆

Addison-Stavlo commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -79,64 +80,92 @@ export default function Header( {
 	const displayBlockToolbar =
 		! isLargeViewport || deviceType !== 'Desktop' || hasFixedToolbar;
 
+	// Testing logic for hovered name...
```
would expect to move this logic into the component in #25320

noahtallen commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1650,6 +1651,18 @@ export function highlightedBlock( state, action ) {
+			return [ action.clientId, ...state ];
+
+		case 'HOVERED_BLOCKS_REMOVE':
+			return filter( state, ( blockId ) => blockId !== action.clientId );
```
some ideas:
- if we need to keep track of multiple, what about using a plain object (e.g. mapping `clientId => true`)?
- is the reason why we include multiple blocks (instead of tracking a single clientId) to make sure that parent blocks are also considered "hovered" when the children are also hovered?

Addison-Stavlo replied:
> is the reason why we include multiple blocks (instead of tracking a single clientId) to make sure that parent blocks are also considered "hovered" when the children are also hovered?

Partially, yeah.  I haven't figured out a good way to track 1 block.  Basically nested blocks create the issue, if we are setting the id onMouseEnter and removing it onMouseLeave, then mousing into nested blocks would set blocks appropriately, but mousing out of nested blocks would clear the state entirely when things are still actually hovered.

Also, if we tracked a single clientId, we would need to bubble up through its parents to see if any are a template part.  However, getBlockParents returns an empty array for blocks at the top level inside a template part, and they don't actually return the template part block as a parent.  So instead we just keep track of all of the blocks that are currently hovered and go through that list to see if any match what we are looking for.

> if we need to keep track of multiple, what about using a plain object (e.g. mapping clientId => true)?
Yeah possibly.  there could still be some benefit to keeping the order they are added, but that could be handled with an array in that object as well or something.

noahtallen replied:
`getBlockParents` ah.... that is probably "unexpected" behavior around the controlled inner blocks code. Well. Kind of expected, but not really. The idea is that the "template part block" is not actually the parent of those blocks _outside_ of the block-editor (e.g. from the point of view of the template part entity). So that's why some of this stuff doesn't work as expected.

noahtallen commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -79,64 +80,92 @@ export default function Header( {
+	);
+	const nameToDisplay = useMemo( () => {
+		// Go through hovered blocks and see if one is of interest.
+		const templatePartBlockId = hoveredBlockIds.find(
```
I wonder if we should make this a selector like `isBlockOfTypeHovered` or something. Maybe premature optimization though

It also seems like it might be expensive to run this filter whenever the hovered block state changes. I wonder if that could be optimized at all 🤔

Addison-Stavlo replied:
> It also seems like it might be expensive to run this filter whenever the hovered block state changes. I wonder if that could be optimized at all 🤔

I agree, that was my initial concern with this approach.  Alternatively, using a different store to just track the name we should display and only adding the event listeners to the blocks we care about would run a lot less often.

noahtallen reviewed (commented):
pretty cool stuff! I'm excited to see where this takes us.

Addison-Stavlo commented:
Pausing this here briefly to implement a rough idea of another approach so we can compare before moving forward.

Addison-Stavlo committed:
remove ugly header and div wrapper

Addison-Stavlo committed:
Merge branch 'master' into try/hovered-block-tracking

Addison-Stavlo committed:
add listener to wrapper ref, hovered label to new component

Addison-Stavlo committed:
remove added newline

Addison-Stavlo commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -265,6 +268,23 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
+			removeHoveredBlock( clientId );
+		}
+
+		ref.current.addEventListener( 'mouseenter', onMouseEnter );
+		ref.current.addEventListener( 'mouseleave', onMouseLeave );
```
@youknowriad - RE (https://github.com/WordPress/gutenberg/issues/25469#issuecomment-695990845):
> About the block inserter disruption, I wonder if this is related to the use of React events instead of DOM events (event bubbling is distrupted), so I'd consider trying event handlers using the "ref" in the BlockWrapper component.

I think this is what you were proposing?  It still seems to have the same issue, as the `mouseleave` fires for a block when we hover over an inserter inside it.

One way to get around this would be to replace add/removeHoveredBlock with a `setHoveredBlocks` in both cases.  We could use `elementsFromPoint` to get the elements, make a list of the clientIds listed on those elements, and pass that list to `setHoveredBlocks`. 🤔

youknowriad replied:
I see. I don't have a specific idea, worth trying yours. One important thing to check is the performance impact for long posts when you're moving the mouse in the editor.

Addison-Stavlo replied:
Updated to try what I was thinking. It seems to behave well.

Addison-Stavlo committed:
small refactor

Addison-Stavlo committed:
rebase goo

Addison-Stavlo committed:
try setHoveredBlocks with elements under cursor

Addison-Stavlo committed:
use the getBlockDisplayText function

Addison-Stavlo committed:
make useSelect destructuring consistent

Addison-Stavlo committed:
remove unused filter

Addison-Stavlo commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -265,6 +265,36 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
 	}, [ isNavigationMode, isHovered, setHovered ] );
 
+	useEffect( () => {
+		function getHoveredBlocksFromCursor( event ) {
+			const hoveredElements = document.elementsFromPoint(
+				event.clientX,
+				event.clientY
+			);
+			const blockIds = [];
+			hoveredElements.forEach( ( element ) => {
+				if ( element.dataset.block ) {
+					blockIds.push( element.dataset.block );
+				}
+			} );
+			return blockIds;
+		}
+		function onMouseEnter( event ) {
+			setHoveredBlocks( getHoveredBlocksFromCursor( event ) );
+		}
+		function onMouseLeave( event ) {
+			setHoveredBlocks( getHoveredBlocksFromCursor( event ) );
+		}
```
This is our way of working around the inserter disruption (https://github.com/WordPress/gutenberg/pull/25348/files#r492402219).  Whenever a hover (enter/leave) fires, we parse the blocks under the cursor to retain the fidelity of the hovered blocks list.  

I don't currently see this adding performance issues, the worst case would be having a lot of nested blocks.  However, setting up ~22 nested blocks at 6x processor slowdown already has poor performance on master and running this PR I don't notice any observable impact to that.

Something I am uncertain about using this approach is how it will behave when we try iframe-ing editor content, depending on how that is set up and if this is trying to parse through nested iframes. 🤔

jeyip replied:
> However, setting up ~22 nested blocks at 6x processor slowdown already has poor performance on master and running this PR I don't notice any observable impact to that.

I can second this. I played around with non-nested content with 15+ paragraph blocks, 10+ images, 10+ template parts, and, generally speaking, the performance seemed fine. 

With 4x and 6x slowdown, relative to the performance of the existing editor on master, there wasn't an observable impact on interaction latency. To be fair though, the editor is already pretty slow when on 4x and 6x slowdown.

Addison-Stavlo reviewed (commented):
Could use some eyes and thoughts on this.  Will update jsdocs/tests/comments etc. if we solidify that this is the way to go.

jeyip reviewed (commented):
I'm testing hovering on:

- Existing template parts
- Nested template parts
- Newly added template parts

I'm having trouble getting the functionality to work in IE11 (hovering template parts doesn't change the topbar label), but it feels awesome in Firefox, Chrome, Edge, and Safari.

jeyip commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1650,6 +1650,15 @@ export function highlightedBlock( state, action ) {
 
+export function hoveredBlocks( state = [], action ) {
+	switch ( action.type ) {
+		case 'HOVERED_BLOCKS_SET':
```
`Non-blocking`:

This name gave me pause. It looks like most other action types use the convention of VERB_NOUN (Ex. SELECT_BLOCK, SET_NAVIGATION_MODE, etc.) Is there a reason we don't do the same here? 

The other reason I bring this up is because I thought we were describing a "set" in the sense of the [abstract data type](https://en.wikipedia.org/wiki/Set_(abstract_data_type)), and I spent a while wondering why state management wasn't implemented with [a javascript set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set).

Addison-Stavlo replied:
good call, updated!

jeyip reviewed (commented):
I really like the solution you propose to solve the inserter bug ⭐ 

It matches previously established patterns in the `block-wrapper` component (with refs and event listeners), it's easy to follow in the source code, and the performance you've described is promising. I'm still playing around with performance locally, but the PR is looking good to me.

Addison-Stavlo committed:
name action type by standard convention

Addison-Stavlo commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1650,6 +1650,15 @@ export function highlightedBlock( state, action ) {
+export function hoveredBlocks( state = [], action ) {
+	switch ( action.type ) {
+		case 'SET_HOVERED_BLOCKS':
+			return action.hoveredBlockIds;
```
I wonder if we should make this an array of the actual block objects as opposed to just the clientIds.  And if so I wonder if we should do that here or in the selector. 🤔   I don't think it makes a big difference for our needs, but maybe this would be more useful if it contained more information so people wouldn't need to use `getBlock` or `getBlockName` to analyze it. 🤷‍♀️

noahtallen replied:
i think the clientIds make sense because when block change, these might be stale.

Addison-Stavlo replied:
True, but that wouldn't be stale if we retrieved the blocks from the ids when the selector runs?  either way its probably not necessary.

noahtallen commented on packages/edit-site/src/components/header/document-actions/index.js:
```diff
@@ -25,16 +25,41 @@ function useSecondaryText() {
+			label: getBlockDisplayText(
+				getBlock( hoveredTemplatePartBlockId )
+			),
+			isActive: true,
```
Not sure exactly where the designs will take this, but my thought was that `isActive` could be false in this case

Addison-Stavlo replied:
Oh interesting thought, I totally missed that idea.  So the hovered item will be inactive.  We can do something like this, but I think we would need to complicate the code in this `useSecondayText` hook a bit.  Currently, hover takes priority over selection, so if we have a template part (or child) selected, it will read as non-active if the cursor is over it as well (if we are returning `isActive: false` here).  So we will need to restructure the hook with that in mind, and return `isActive` in the hover state depending on whether or not the hovered item we want to show corresponds to the currently selected one.  Perhaps a good refactoring for a follow up?

noahtallen replied:
Sure!

Addison-Stavlo replied:
>Perhaps a good refactoring for a follow up?

Or not 🤣  ... just refactored to handle this in - 5548826 .  So the hovered item will be inactive if it doesn't correspond to the selected one.

noahtallen reviewed (commented):
this is looking pretty cool! cc @shaunandrews to see how this feels.

I definitely like it a lot. Have we tried running the performance benchmarks locally?

Addison-Stavlo commented on packages/edit-site/src/components/header/document-actions/index.js:
```diff
@@ -25,16 +25,41 @@ function useSecondaryText() {
 	} );
 
+	// Go through hovered blocks and see if one is of interest.
+	const hoveredTemplatePartBlockId = hoveredBlockIds.find(
```
Maybe!  I don't think we have an immediate need for it, but it could be helpful for future use.

To be clear... do you mean something like `isBlockHovered( block.name )` or `isBlockHovered( clientId )`.  The second one is what makes sense to me, but from the context of the line you pinned this not on Im thinking you might mean the first?

noahtallen replied:
I think both could be helpful for different use cases, but the first is obviously more helpful to us here! Maybe `isBlockOfTypeHovered` or something more generic... I like the idea of adding it as a selector because it helps separate concerns a bit with this selector :)

Addison-Stavlo replied:
> the first is obviously more helpful to us here!

Thinking about it more Im not totally following.  Is there a need to retrieve a boolean for `is` one of that type hovered?  Or just to retrieve it if it is, like `getHoveredBlockByType` (we could use something like that here)?  I guess there are a handful of things that could potentially be useful, but does it make sense to add them all proactively or as we need them?

Addison-Stavlo replied:
maybe something like 6f3751e ?

Addison-Stavlo committed:
refactor useSecondaryText to determine if hovered item should be active

shaunandrews commented:
I noticed that the template part name appears to overflow the top bar, and doesn't fade-in:

![document-info-animate-bug](https://user-images.githubusercontent.com/191598/94184334-c0586700-fe71-11ea-8606-6912827272a6.gif)

--

Here's how I'd expect the transitions to look:

![document-info-animation](https://user-images.githubusercontent.com/191598/94184845-85a2fe80-fe72-11ea-96ae-84c6e81571c1.gif)

Here's the same animation slowed down to help see the transitions:

![document-info-animation-slow](https://user-images.githubusercontent.com/191598/94185257-124dbc80-fe73-11ea-9e69-b5cbcdb6dccb.gif)


Hovering:
- Template name (bold, grey-900) transitions up
- Template part name (normal, grey-700) fades into view and transitions up

Selecting:
- Template name goes transitions to normal font-weight and grey-700
- Template part name transitions to bold and grey-900

Addison-Stavlo commented:
Thanks Shaun!  

Ive linked your comments regarding the animation to https://github.com/WordPress/gutenberg/issues/25545 so we can follow up on polishing that.  Here, we aren't editing the animation, but adding a system which allows us to show the name of the hovered Template Part.

Addison-Stavlo committed:
check event timeStamp before evaluating

Addison-Stavlo commented:
> Have we tried running the performance benchmarks locally?

For the `test-performance` script, all tests pass and there is no noticeable difference in the output (listed times were oddly better on this branch than master, so whatever difference there is seems to be lost in the noise regarding that).
<img src="https://user-images.githubusercontent.com/28742426/94209234-c9f5c500-fe99-11ea-8323-c4b6df1c287e.png" width="400" />

However, I looked into my manual testing some performance when hovering.  To set up I nested a paragraph block inside 21 (otherwise empty) group blocks, and set the processor to throttle at a 6x slowdown.  I then tested hovering over and away from the paragraph block.

I did find that the handler firing for all 22 blocks simultaneously did cause a large difference in old vs new hover performance:
![Screen Shot 2020-09-24 at 6 42 27 PM](https://user-images.githubusercontent.com/28742426/94208809-b6962a00-fe98-11ea-8a6b-eecb81bd7777.png)
Where the first block in that above image is for the existing `mouseover` event, while all the `mouseenter` duplicates are a result of what was added here.

I added a step to check/track the `event.timeStamp` to prevent this from happening.  So running that test again then gave me these results (with no duplicates fired):
![Screen Shot 2020-09-24 at 6 13 05 PM](https://user-images.githubusercontent.com/28742426/94209056-5653b800-fe99-11ea-8a12-0913ee0d74a5.png)
Where the 113ms corresponds to the existing event and the 67ms corresponds to what we added here (for hovering 22 nested blocks at 6x processor slowdown).

Addison-Stavlo committed:
add comment to timeStamp check

Addison-Stavlo committed:
add getHoveredBlockByBlockName selector

Addison-Stavlo committed:
don't pass an object to the dispatch

Addison-Stavlo committed:
add comment for parsing elements under cursor

noahtallen commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -265,6 +269,50 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
+			// Check the timeStamp of the last time this was set.
+			// This prevents needlessly rerunning the parse and dispatch for nested blocks
+			// who share boundaries that trigger the mouse events at the same time.
+			if ( event.timeStamp !== getHoveredBlocksTimeStamp() ) {
```
Would it make sense to store this timestamp locally (maybe with `useRef`?) instead of having to send it to the store?

Addison-Stavlo replied:
Wouldn't that still create a different ref for each instance of the wrapper?

Addison-Stavlo replied:
Id like to keep it out of the store as well, but I wasn't quite sure how to make that work. 🤔

noahtallen replied:
> Wouldn't that still create a different ref for each instance of the wrapper

ahh good point. I guess it would have to be at a higher level, or in the global file scope. Global file scope would be really easy to do, but it doesn't feel "best practice" I guess

Addison-Stavlo replied:
🤔   Yeah... Im not sure which way is better, I'm open to either direction.  While I am not a huge fan of complicating the store with it, it is optional and doesn't necessarily need to be used by anyone else ever setting the values.

jeyip replied:
While I'm also wary of adding this to the store, I see this as a legitimate use case. The only other alternatives I can think of are using context or using local state in a common parent component and prop drilling (`packages/edit-site/src/components/block-editor/index.js` maybe?)

I'd also prefer to avoid global file scope for state if only because it subverts expectations and its another pattern to keep track of when handling state management for components.

Addison-Stavlo replied:
I wondered about prop drilling as well, it seems like it may be a bit more ugly.  There may be a better way than I am thinking, but it would need to be passed into `useBlockWrapperProps` everywhere it is used (which includes a lot of edit.js block files).  I think the store may be the cleanest home for the timestamp.

> While I'm also wary of adding this to the store, I see the timestamp as a legitimate use case. It's information that needs to be shared amongst many components.

Very true.

Addison-Stavlo commented:
@youknowriad any thoughts on where this has gone?  

If it seems like something we could try out maybe we should gate adding the event handlers behind the FSE experiment being active (and labeling the selectors/dispatch experimental)?

noahtallen commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1798,3 +1798,18 @@ export function isBlockHighlighted( state, clientId ) {
+	const matchingBlockClientId = state.hoveredBlocks.hoveredBlockIds.find(
+		( blockId ) => getBlockName( state, blockId ) === blockName
+	);
+	return getBlock( state, matchingBlockClientId );
```
I think it would be good to memoize this selector (maybe on `state.hoveredBlocks.hoveredBlockIds` and `state.blocks.cache`)

Addison-Stavlo replied:
✅  set up the memoize with `createSelector`

noahtallen commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1671,4 +1683,5 @@ export default combineReducers( {
 	hasBlockMovingClientId,
 	automaticChangeStatus,
 	highlightedBlock,
+	hoveredBlocks,
```
I wonder if this should be added under the `blocks` reducers. Maybe not since it isn't crucial for blocks 🤔

Addison-Stavlo replied:
Im not certain, but I lean on the "maybe not" as well since it isn't crucial and is more experimental.  🤷‍♀️

noahtallen commented on packages/edit-site/src/components/header/document-actions/index.js:
```diff
@@ -25,45 +25,62 @@ function useSecondaryText() {
-			label: selectedBlockLabel,
-			isActive: true,
+			label: hoveredBlockLabel,
+			isActive: hoveredBlockLabel === selectedBlockLabel,
```
I think we should use client ID to compare. I could see a case where two labels might be the same (even though it'd be confusing)

Addison-Stavlo replied:
✅  Set it up to compare by clientId

noahtallen reviewed (commented):
This works wonderfully for me locally. I left some comments about nitpicky code stuff. I also agree about making the selectors experimental. I like the concept of turning off the onmouseover etc handlers outside of FSE for now (so we can make sure it works well), but I'm not sure how straightforward that would be.

Addison-Stavlo commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1798,3 +1798,18 @@ export function isBlockHighlighted( state, clientId ) {
+	return state.hoveredBlocks.hoveredBlockIds;
+}
+
+export function getHoveredBlockByBlockName( state, blockName ) {
```
definitely!  I was holding off to make sure we were on the right track, but its probably time to add some doc comments.

Addison-Stavlo replied:
comments added

Addison-Stavlo committed:
restrict event handlers to FSE experiment

Addison-Stavlo committed:
oops, restrict everything in the effect to FSE experiment

Addison-Stavlo committed:
convert to __experimental labels + memoize selector

Addison-Stavlo committed:
hovered label comparison for active state by clientId

Addison-Stavlo committed:
limit reruns to 10ms bc firefox doesnt create timestamps right

noahtallen commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1798,3 +1798,45 @@ export function isBlockHighlighted( state, clientId ) {
+		);
+		return getBlock( state, matchingBlockClientId );
+	},
+	( state ) => [ state.hoveredBlocks.hoveredBlockIds, state.blocks.cache ]
```
I didn't realize this (so my bad for the recommendation), but I now realize that `state.blocks.cache` won't work. It's meant to be like `state.blocks.cache[ clientId ]`, which contains an object which only changes reference when the block is updated. Given that we don't have a clientId here, I guess it doesn't make sense to memoize on that.

But... If we only memoize on hoveredBlockIds, it would return stale block attributes if the block attributes have updated but the hovered state has not updated. 

So maybe we should only return the client ID, and then only memoize on `state.hoveredBlocks.hoveredBlockIds`? (I don't think block name can change at runtime)

Hm.

Addison-Stavlo replied:
So I just did these changes locally, then started wondering more about:
> I now realize that `state.blocks.cache` won't work.

At first I would think this is true because we are checking the same object.  But looking at the cache reducer, every action that changes the cache, creates a new cache object and old values are spread into it.  So any time the cache is updated, `state.blocks.cache` changes to a new object, and it should work as a dependency here? 🤔

noahtallen replied:
that's true. I guess that would refresh the selector whenever any block changes which makes sure the returned block is up to date :)

I still like the idea of just returning the clientId and letting it just cache on hovereBlocks. Feels more "pure" that way, I guess.

Either way, it's not a big deal IMO.

Addison-Stavlo replied:
true!  and I have that all ready to go, might as well. 😄

noahtallen reviewed (commented):
Thanks for adding those updates! I added a note about something I didn't realize about the memoization.

Addison-Stavlo commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -265,6 +275,55 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
+				// Check the timeStamp of the last time this was set.
+				// This prevents needlessly rerunning the parse and dispatch for nested blocks
+				// who share boundaries that trigger the mouse events at or around the same time.
+				if (
+					event.timeStamp - ( getHoveredBlocksTimeStamp() || 0 ) >
+					10
+				) {
```
So this is bugging me.  Of course different browsers can't handle events consistently 🤦‍♀️  .

So Chrome will make `event.timeStamp` the same for all events created at the same time.  Firefox handles the first event fully and then creates the second one, so the timeStamps end up being later based on how long the first took to run.  In practice with running the parse and dispatch, thats ~5ms.  Additionally, Firefox dev tools doesn't seem to have a way to CPU throttle, so testing this for slower CPUs isn't straight forward in that browser.  

So if we want this fix for not unnecessarily running the evaluation to work on firefox we need to do `timeStamp1 - teimpStamp2 > someDelta` instead of `timeStamp1 !== timeStamp2`.  But from what I saw previously (in testing on chrome) this parse can take between 40-70ms to run on a slow (x6 slowdown processor).  If we make this delta something like 70-100ms, we start experiencing funky behavior when moving the cursor out of blockA and into blockB sometimes not re-evaluating what is hovered (although still probably better than the inserter disrupting the hover state).

🤷‍♀️  😭  ❓ 😠  not sure what to think at this point.

Addison-Stavlo replied:
creating a new timeStamp just before the dispatch may help, I will evaluate this a bit more tomorrow.

Addison-Stavlo replied:
Yep, this seems to take care of things.

Addison-Stavlo committed:
create a new timeStamp after evaluation instead of using the event's original one

Addison-Stavlo commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -265,6 +275,55 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
 	}, [ isNavigationMode, isHovered, setHovered ] );
 
+	useEffect( () => {
+		if ( isFullSiteEditingActive ) {
```
Yeah... hooks can be mean with their rules! 😆

Addison-Stavlo commented:
Im wondering if it would make more sense to limit these event handlers to the relevant block types only (so just template-part for now).  🤔   While this current solution should work well with the hoops we have jumped through to help performance on poor systems as much as possible, we don't actually have a need to track hover ALL block types (just a very small subset).  While it would be nice to have a general system for being able to track all hovered blocks, limiting the handlers to only those wrappers could allow us to get rid of the worry and all the extra complexity requiring timestamps and limiting unnecessary reruns of the logic.

noahtallen commented:
> While it would be nice to have a general system for being able to track all hovered blocks

Yeah, good question. I guess to me, we'd need to change the selector to be more like "getHoveredTemplatePart", which seems very strange for the block-editor store by itself. At the very least, I think we would need a way to register which block types to look out for. And at that point, I feel like we're just adding extra complexity just to restrict things without a huge reason for it. As a plugin developer, it's always very nice to just have information like this available without extra work (like currently selected block)

if this isn't introducing big performance issues with the timestamp comparison, I think it's fine to just go for tracking all of them

Addison-Stavlo commented:
>As a plugin developer, it's always very nice to just have information like this available without extra work (like currently selected block)
if this isn't introducing big performance issues with the timestamp comparison, I think it's fine to just go for tracking all of them

Good points, thanks!

Addison-Stavlo committed:
update selector to return Id instead of block

Addison-Stavlo committed:
reduce lockout timer to what is necessary

Addison-Stavlo committed:
nevermind that

david-szabo97 commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -265,6 +275,60 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
 	}, [ isNavigationMode, isHovered, setHovered ] );
 
+	useEffect( () => {
+		if ( isFullSiteEditingActive ) {
```
How about an early return? We can save an indentation 😋 
```js
if ( ! isFullSiteEditingActive ) {
	return;
}
```

Addison-Stavlo replied:
Thats definite better. 😀

david-szabo97 commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -265,6 +275,60 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
+				} );
+				return blockIds;
+			}
+			async function evaluateHoveredBlocks( event ) {
```
Any reason for this to be `async`?

david-szabo97 commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -265,6 +275,60 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
+			// To accurately determine which blocks are hovered we must look at the elements under the cursor.
+			// Internal block inserters will fire events like mouseleave for blocks they are visually contained within,
+			// which makes an add/remove by clientId per individual block approach inaccurate.
+			function getHoveredBlocksFromCursor( event ) {
```
Can move this outside of the `useBlockWrapperProps`? I don't see any references that could stop us from doing it. Though it may feel weird if it's moved from here because it makes it less easier to read. We need to lookup the function.

Addison-Stavlo replied:
Whats the thought with that move?  That it could be exported and used elsewhere?  Im open to the idea of moving, although Im not exactly sure where to move it and not very convinced it should.  But Id also think if we were to move it we should tie in everything with the timestamps/whatever else we need to limit the unnecessary recalculations? 🤔

david-szabo97 replied:
It's my personal preference, but usually I try to avoid declaring functions inside functions if the inner function isn't using any variables/references from the outer scope. We can make the `useBlockWrapperProps` a bit shorter and less noisy.

Also, if we move `getHoveredBlocksFromCursor` function outside, then we will have only one reference to this function, rather than one reference to each `useBlockWrapperProps` call.

TLDR: Micro optimization and personal preference :D 

`evaluateHoveredBlocks` is tied to the outer scope so we can't move that though.

Addison-Stavlo replied:
Yeah i definitely agree.  I moved both functions to another file to clean it up a bit.

Addison-Stavlo committed:
convert to early return and rid of unnecessary async func

Addison-Stavlo commented:
Instead of limiting by timestamps, a better option could be to limit on mouse position of the last evaluation.  While simultaneous events may have slightly different timestamps in some browsers, it seems the mouse positions are always the same.

However, we would need to account for scrolling.  Otherwise, when the mouse is in the same position and we scroll down the document it would never re-evaluate what is hovered.

`event.pageX` and `event.pageY` would normally work, however since the editor and its scrollbar are internal to the actual document this essentially works the same as `event.clientX` / `event.clientY`.  Similarly, we cannot combine this with scroll location like `window.scrollY` since the scrollbar in the editor is internal to the document.

We could:
* do a combination of pointer location and timestamp - Only re-evaluate if the pointer position is different OR some deltaT has passed since the last calculation.
* Find a non-hacky way of determining the editors scroll location and combine this with pointer location (no timestamp necessary).
* Just limit on pointer location alone and be fine with the fact that the hovered entities wont update scrolling down the page until the mouse is moved again.

Number 2 would be good if there is already a way to do this. 🤔   Kind of a shame pageX/pageY won't work on their own.

jeyip commented:
> Find a non-hacky way of determining the editors scroll location and combine this with pointer location
To clarify, I'm not the most familiar with scroll attributes, but couldn't evaluate `scrollTop` and `scrollLeft` on the editor?

To be more specific, I tested this in the chrome dev console. `document.querySelector('div.interface-interface-skeleton__content').scrollTop`

Addison-Stavlo committed:
Merge branch 'master' into try/hovered-block-tracking

Addison-Stavlo commented:
>  but couldn't we evaluate scrollTop and scrollLeft on the editor?

Yeah we could do that.  I was hoping there was a better way to get ahold of that info without `querySelector` but haven't been having much luck.  That interface skeleton is from a different package so it might be a bit odd to have something from the block-editor rely on it as well.

Addison-Stavlo committed:
compare previous mouse coords

Addison-Stavlo commented:
> do a combination of pointer location and timestamp - Only re-evaluate if the pointer position is different OR some deltaT has passed since the last calculation.

Tried this for now.  Although it would be nice to get rid of that timestamp, we will need some way to detect scroll position in combination of mouseCoords for that.  The timestamp is now just used to allow scrolling over blocks with the cursor in the same position to re-trigger the evaluation.

Addison-Stavlo committed:
move hovered blocks functions to a separate file

Addison-Stavlo committed:
update docs and comments

Addison-Stavlo commented on packages/block-editor/src/components/block-list/hovered-blocks.js:
```diff
@@ -0,0 +1,64 @@
+export function evaluateHoveredBlocks( event ) {
+	// Check last event data to determine if we should reevaluate hovered blocks.
+	// Requiring different mouseCoords limits unnecessary evaluations triggered from the same
+	// event firing from multiple nested blocks.  Checking the time allows us to re-evaluate
+	// even when mouseCoords have not changed, such as when scrolling via mouse-wheel.
+	// However, this deltaTime checked must be sufficiently long to ensure slow CPUs do not
+	// trigger unnecessary recalculations for events triggered simultaneously.
+	const { time, mouseCoords } =
```
Scroll location would be much better to check than the last timeStamp to ensure this works on scrolling as well.  However, the scroll location corresponds to the `interface-interface-skeleteton__content` which is contained in the interface package.  Whether we found a way to reference this element or querySelector'd for it here, we would make this functionality dependent on a separate package which is also not ideal.  The time check will work, but fast mouse-wheel scrolls may not always re-trigger the evaluation.

Addison-Stavlo commented:
Im thinking this should be more or less 'ready to go' at this point.

noahtallen reviewed (approved):
I'm in favor of merging this to try it out. The code looks good, and it tests well locally. Thanks for all the iterations @Addison-Stavlo!

I don't think we should spend any more time optimizing it though. So if folks don't agree that we should merge and try out, I think we should move on from this for now.

Addison-Stavlo committed:
Merge branch 'master' into try/hovered-block-tracking

Addison-Stavlo committed:
rebase and fix merge conflicts

vindl reviewed (approved):
I'm also in favor of trying this out, we can always revise the approach later if we come up with a better way to resolve the underlying issue. It tests well for me and the code is looking good. Nice work @Addison-Stavlo! ✨

youknowriad commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1645,6 +1645,33 @@ export function highlightedBlock( state, action ) {
+			const { hoveredBlockIds, timeStamp, mouseCoords } = action;
+			return {
+				hoveredBlockIds,
+				eventData: {
```
Why do we need thisi information in state?

noahtallen replied:
the scenario is: "move mouse over a block which is deeply nested". This will trigger N hover events for N nested blocks. If we have a ton of those blocks, we'll be dispatching to the store N times, which caused performance issues. To avoid updating the store that often, we need to compare the timestamp of an event with the timestamp of another event. So we track the timestamp here. I think the other option would be storing it in a global variable in the file where it's used

youknowriad replied:
Is there a way to "stop propagation" somehow on the event?

noahtallen replied:
i'm unsure, I'll leave that for @Addison-Stavlo to answer

Addison-Stavlo replied:
> Is there a way to "stop propagation" somehow on the event?

I didn't have any luck.  It's not one event bubbling, but many events being triggered simultaneously.

I think we can get around requiring this information and extra complexity by debouncing. e41e82d

Addison-Stavlo replied:
That said, while having a way to track which blocks are hovered is neat, given the comments from design below it doesn't seem like we have a consensus on where we should use this information yet (probably not in the document actions header from the sounds of it 😆 ).  This could prove useful in implementing something similar in the future (such as closer to the cursor https://github.com/WordPress/gutenberg/pull/25348#issuecomment-710303993), but until we figure out the right place to utilize this there is no rush for adding the handlers and store either.

noahtallen commented on packages/block-editor/src/components/block-list/block-wrapper.js:
```diff
@@ -87,10 +90,13 @@ export function useBlockProps( props = {}, { __unstableIsHtml } = {} ) {
 					: undefined,
 				isNavigationMode: _isNavigationMode,
+				isFullSiteEditingActive: getSettings()
+					.__experimentalEnableFullSiteEditing,
```
since this feature is only used by FSE right now, we thought it might be wise to only enable it in FSE to avoid any possible performance regressions in the main plugin

youknowriad replied:
Did we check/notice performance issues?

noahtallen replied:
yeah, we did spend time looking at and fixing performance problems.

youknowriad replied:
So just to confirm, without this check we do have performance issues?

youknowriad replied:
What I'm saying is that performance doesn't seem to be a valid reason to add this check because otherwise FSE screen won't be performant anyway right?

noahtallen replied:
> So just to confirm, without this check we do have performance issues?

none that we have uncovered in the testing. We aren't gating this because of any performance issue we know of, just out of an abundance of caution

youknowriad replied:
I think personally we should remove it in that case, maybe add a comment explaining that it's only useful for FSE for now but i don't think we should be sacrificing puriness and architecture in favor of performance issues we're not sure exist.

shaunandrews commented:
I have mixed feelings here. One side of me likes the interactivity and the ability to quickly see what template parts make up my document; But the other side of me things maybe there's too much changing as I move my mouse pointer around the screen.

As is, the transition doesn't handle nested template parts well; I'd expect the label to animate even for nested template parts.

--

I think for the normal edit mode, this might be too much motion on the screen. However, maybe it could make sense as part of select mode, which already shows block borders on hover.

jameskoster commented:
I'd agree that the visual feedback can be helpful, but if we want to add something I'd like for it to be closer to the cursor. Trying to mouse over my Footer template part, but having to look at the top of my screen to understand when I've hit the target is kind of uncomfortable.

Addison-Stavlo committed:
use debounce instead

Addison-Stavlo committed:
Merge branch 'master' into try/hovered-block-tracking

Addison-Stavlo commented:
a bit old and stale, past due for a close button click.
